pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -306,15 +308,336 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
 	return result, nil
 }
 
-// SyncPod syncs the running pod into the desired pod.
-func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus,
-	podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret,
-	backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
-	result.Fail(fmt.Errorf("not implemented"))
+// containerToKillInfo contains neccessary information to kill a container.
+type containerToKillInfo struct {
+	// The spec of the container.
+	container *api.Container
+	// The name of the container.
+	name string
+	// The message indicates why the container will be killed.
+	message string
+}
+
+// podContainerSpecChanges keeps information on changes that need to happen for a pod.
+type podContainerSpecChanges struct {
+	// Whether need to create a new sandbox.
+	CreateSandbox bool
+	// The id of existing sandbox. It is used for starting containers in ContainersToStart.
+	SandboxID string
+	// The attempt number of creating sandboxes for the pod.
+	Attempt uint32
+
+	// ContainersToStart keeps a map of containers that need to be started, note that
+	// the key is index of the container inside pod.Spec.Containers, while
+	// the value is a message indicates why the container needs to start.
+	ContainersToStart map[int]string
+	// ContainersToKeep keeps a map of containers that need to be kept as is, note that
+	// the key is the container ID of the container, while
+	// the value is index of the container inside pod.Spec.Containers.
+	ContainersToKeep map[kubecontainer.ContainerID]int
+	// ContainersToKill keeps a map of containers that need to be killed, note that
+	// the key is the container ID of the container, while
+	// the value contains neccessary information to kill a container.
+	ContainersToKill map[kubecontainer.ContainerID]containerToKillInfo
+}
+
+// podSandboxChanged checks whether the spec of the pod is changed and returns
+// (changed, new attempt, original sandboxID if exist).
+func (m *kubeGenericRuntimeManager) podSandboxChanged(pod *api.Pod, podStatus *kubecontainer.PodStatus) (changed bool, attempt uint32, sandboxID string) {
+	if len(podStatus.SandboxStatuses) == 0 {
+		return true, 0, ""
+	}
+
+	readySandboxCount := 0
+	for _, s := range podStatus.SandboxStatuses {
+		if s.GetState() == runtimeApi.PodSandBoxState_READY {
+			readySandboxCount++
+		}
+	}
+
+	// Needs to create a new sandbox when readySandboxCount > 1 or the ready sandbox is not the latest one.
```

Comment:
If readySandboxCount > 1, then it's hard to distinguish which are the unwanted sandboxes and whether the containers are running at right sandbox. So just kill all and create a new one.